File: 1D_DP/C/UniqueBST.c
#include<stdio.h>


int uBST(int n, int *a){
    if(n <= 1) return 1;

    if(a[n]) return a[n];

    for(int i = 1 ; i <= n ; i++){
        a[n] += uBST(i - 1,a) * uBST(n - i,a); 
    }

    return a[n];


}

int main(){
    int n = 3;
    int dp[4] = {0 ,0 ,0, 0};
    int *a = dp;
    printf("Unique No of BST are %d",uBST(n,a));
    return 0;
}

File: Dynamic _Programming/1D_DP/PYTHON/UniqueBST.py
def uBST(n,dp):
    if n <= 1:
        return 1
    
    if dp[n] != 0:
        return dp[n]
    
    for i in range(1,n+1):
        dp[n] += uBST(i - 1,dp) * uBST(n - i,dp)

    return dp[n]

n = 5
dp = [0] * (n+1)
print(uBST(n,dp))    